File: backend/src/app/routers/memory.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.db import get_session
from app.knowledge.loader import get_graph
from app.models.memory import UserProfile
from app.routers.deps import current_user_id
from app.schemas.memory import MasteryItem, MemorySnapshot
from app.services import memory_service

router = APIRouter(prefix="/memory", tags=["memory"])


@router.get("", response_model=MemorySnapshot)
async def get_memory(
    user_id: uuid.UUID = Depends(current_user_id),
    session: AsyncSession = Depends(get_session),
) -> MemorySnapshot:
    graph = get_graph()
    items = await memory_service.snapshot(session, user_id)
    weak = await memory_service.weak_concepts(session, user_id)
    profile = await session.get(UserProfile, user_id)
    return MemorySnapshot(
        items=[
            MasteryItem(
                concept_id=m.concept_id,
                concept_name=graph[m.concept_id].name if m.concept_id in graph else m.concept_id,
                confidence=m.confidence,
                correct_count=m.correct_count,
                incorrect_count=m.incorrect_count,
                last_seen_at=m.last_seen_at,
            )
            for m in items
        ],
        weak=[m.concept_id for m in weak],
        preferred_style=(profile.preferred_style if profile else "mixed"),
    )

File: mobile/lib/models/memory.dart
class MasteryItem {
  MasteryItem({
    required this.conceptId,
    required this.conceptName,
    required this.confidence,
    required this.correctCount,
    required this.incorrectCount,
  });

  final String conceptId;
  final String conceptName;
  final double confidence;
  final int correctCount;
  final int incorrectCount;

  factory MasteryItem.fromJson(Map<String, dynamic> j) => MasteryItem(
        conceptId: j['concept_id'] as String,
        conceptName: j['concept_name'] as String,
        confidence: (j['confidence'] as num).toDouble(),
        correctCount: j['correct_count'] as int,
        incorrectCount: j['incorrect_count'] as int,
      );
}

class MemorySnapshot {
  MemorySnapshot({required this.items, required this.weak, required this.preferredStyle});

  final List<MasteryItem> items;
  final List<String> weak;
  final String preferredStyle;

  factory MemorySnapshot.fromJson(Map<String, dynamic> j) => MemorySnapshot(
        items: (j['items'] as List)
            .map((e) => MasteryItem.fromJson(e as Map<String, dynamic>))
            .toList(),
        weak: (j['weak'] as List).cast<String>(),
        preferredStyle: j['preferred_style'] as String,
      );
}
